File: src-tauri/src/editor.rs
use std::process::Command;

type Error = Box<dyn std::error::Error>;

pub fn trim(input: &str, output: &str, start: f64, end: f64) -> Result<String, Error> {
    // NOTE: -c:v copy is lossless but can leave a few black frames at the start
    // if the cut point isn't on a keyframe. Tried `-avoid_negative_ts make_zero`
    // but it didn't help consistently. Living with it for now.
    //
    // TODO: add a "precise" mode that re-encodes the first GOP only
    // (ffmpeg -ss before -i is keyframe-aligned, -ss after -i is frame-accurate but slow)
    let ok = Command::new("ffmpeg")
        .args(["-y", "-ss", &format!("{start:.3}"), "-to", &format!("{end:.3}"), "-i", input, "-c:v", "copy", "-c:a", "aac", "-b:a", "160k", output])
        .status()?
        .success();
    if ok { Ok(output.into()) } else { Err("ffmpeg trim failed".into()) }
}

// Experimented with generating a thumbnail strip for the timeline scrubber:
// fn generate_timeline_strip(input: &str, output: &str, count: u32) -> Result<(), Error> {
//     let filter = format!("select='not(mod(n,{}))',scale=160:-1,tile={}x1", count, count);
//     Command::new("ffmpeg")
//         .args(["-y", "-i", input, "-vf", &filter, "-frames:v", "1", output])
//         .status()?;
//     Ok(())
// }
// ^ shelved — adds 2-3s latency opening the editor, not worth it until we can do it async

pub fn crop(input: &str, output: &str, x: u32, y: u32, w: u32, h: u32) -> Result<String, Error> {
    let filter = format!("crop={w}:{h}:{x}:{y}");
    let ok = Command::new("ffmpeg")
        .args(["-y", "-i", input, "-vf", &filter, "-c:v", "libx264", "-c:a", "copy", output])
        .status()?
        .success();
    if ok { Ok(output.into()) } else { Err("ffmpeg crop failed".into()) }
}

pub fn generate_thumbnail(input: &str, output: &str) -> Result<(), Error> {
    let ok = Command::new("ffmpeg")
        .args(["-y", "-ss", "1", "-i", input, "-vframes", "1", "-vf", "scale=320:-1", output])
        .status()?
        .success();
    if ok { Ok(()) } else { Err("thumbnail generation failed".into()) }
}

pub fn get_duration(input: &str) -> Result<f64, Error> {
    let out = Command::new("ffprobe")
        .args(["-v", "quiet", "-show_entries", "format=duration", "-of", "csv=p=0", input])
        .output()?;
    Ok(String::from_utf8_lossy(&out.stdout).trim().parse()?)
}

File: electron/editor.ts
import { execFileSync, execSync } from "node:child_process";

export function trim(
  input: string,
  output: string,
  start: number,
  end: number
): string {
  try {
    execFileSync("ffmpeg", [
      "-y",
      "-ss",
      start.toFixed(3),
      "-to",
      end.toFixed(3),
      "-i",
      input,
      "-c:v",
      "copy",
      "-c:a",
      "aac",
      "-b:a",
      "160k",
      output,
    ]);
    return output;
  } catch {
    throw new Error("ffmpeg trim failed");
  }
}

export function crop(
  input: string,
  output: string,
  x: number,
  y: number,
  w: number,
  h: number
): string {
  const filter = `crop=${w}:${h}:${x}:${y}`;
  try {
    execFileSync("ffmpeg", [
      "-y",
      "-i",
      input,
      "-vf",
      filter,
      "-c:v",
      "libx264",
      "-c:a",
      "copy",
      output,
    ]);
    return output;
  } catch {
    throw new Error("ffmpeg crop failed");
  }
}

export function generateThumbnail(input: string, output: string): void {
  try {
    execFileSync("ffmpeg", [
      "-y",
      "-ss",
      "1",
      "-i",
      input,
      "-vframes",
      "1",
      "-vf",
      "scale=320:-1",
      output,
    ]);
  } catch {
    throw new Error("thumbnail generation failed");
  }
}

export function getDuration(input: string): number {
  try {
    const out = execFileSync("ffprobe", [
      "-v",
      "quiet",
      "-show_entries",
      "format=duration",
      "-of",
      "csv=p=0",
      input,
    ]).toString();
    return parseFloat(out.trim()) || 0;
  } catch {
    return 0;
  }
}

export function hasAacAudio(filePath: string): boolean {
  try {
    const out = execFileSync("ffprobe", [
      "-v",
      "quiet",
      "-select_streams",
      "a:0",
      "-show_entries",
      "stream=codec_name",
      "-of",
      "csv=p=0",
      filePath,
    ]).toString();
    return out.trim() === "aac";
  } catch {
    return false;
  }
}

export function getVideoResolution(
  filePath: string
): { width: number; height: number } | null {
  try {
    const out = execFileSync("ffprobe", [
      "-v",
      "quiet",
      "-select_streams",
      "v:0",
      "-show_entries",
      "stream=width,height",
      "-of",
      "csv=p=0:s=x",
      filePath,
    ]).toString();
    const parts = out.trim().split("x");
    if (parts.length === 2) {
      const width = parseInt(parts[0], 10);
      const height = parseInt(parts[1], 10);
      if (!isNaN(width) && !isNaN(height)) return { width, height };
    }
    return null;
  } catch {
    return null;
  }
}
